### ISSUE TITLE
Security: 3.20 Tracking issue

### ISSUE THREAD

**Issue by chayim** [sourcegraph/sourcegraph#13130]
### Plan

TBD

### Availability

Period is from August 24th to September 21st. Please write the days you won't be working and the number of working days for the period.

* @chayim: Off for 3 non-consecutive days

### Tracked issues

@ElizabethStirling

Completed
- [x] (🏁 16 days ago) Auth: explicit session invalidation ([~#1126~](https://github.com/sourcegraph/sourcegraph/issues/1126)) 🔒

@chayim

Completed
- [x] (🏁 26 days ago) Security planning and communication changes (~[#1449](https://github.com/sourcegraph/about/pull/1449)~) :shipit:
- [x] (🏁 4 days ago) ~sourcegraph/security-issues~ ([~#83~](https://github.com/sourcegraph/security-issues/issues/83)) 

@daxmc99

- [ ] Update product documentation ([#12266](https://github.com/sourcegraph/sourcegraph/issues/12266)) 

#### Legend

- 👩 Customer issue
- 🐛 Bug
- 🧶 Technical debt
- 🛠️ [Roadmap](https://docs.google.com/document/d/1cBsE9801DcBF9chZyMnxRdolqM_1c2pPyGQz15QAvYI/edit#heading=h.5nwl5fv52ess)
- 🕵️ [Spike](https://en.wikipedia.org/wiki/Spike_(software_development))
- 🔒 Security issue
- :shipit: Pull Request

**Comment by chayim** [comment]
**Last Week**

Last week, I got some approvals for the RFC for encrypted filesystems. I continued working on RFC196 as part of the 3.20 release, with @daxmc99. I started [defining planning process](https://github.com/sourcegraph/about/pull/1449) for the @sourcegraph/security team, and tried out a couple of experiments to better organize the roadmap, and generally make that better. We kicked off (https://github.com/sourcegraph/security-issues/issues/83)[security issue #83] and signed up our external temporary tester.  With @daxmc99 we planned out a sane approach to encrypting data during initialization.

**This Week**

I'll continue to work on (https://github.com/sourcegraph/security-issues/issues/83)[security issue #83]. @daxmc99 and I should make great progress towards encrypting things at startup, and get very close to completion.  I'll close off the [planning process changes](https://github.com/sourcegraph/about/pull/1449), while reviewing the new ~30 candidates in lever for the [security engineer position](https://about.sourcegraph.com/handbook/engineering/hiring/software-engineer-security). Finally, whatever is necessary to get our tester ready for September 7, is also on my plate.

**Comment by ElizabethStirling** [comment]
**Last Week**
Last week, I was unfortunately out sick!

**This Week**
This week, I'm starting by getting caught up on emails and code reviews. Later, I plan to finish up #1126. Depending on how difficult OAuth authorization is to invalidate, I may need to break this into two separate PRs, one which logs the user out of all session-based authentications when they sign out, and one that handles OAuth invalidation and creates buttons on the frontend for the site admin to invalidate all sessions belonging to a user.

**Comment by chayim** [comment]
**This Week**

Closed off [security issue #83](https://github.com/sourcegraph/security-issues/issues/83) and got an interim report by week's end. Still waiting for the final report.  Also closed off and got approval for [RFC 214](https://docs.google.com/document/d/1ajVQ11Vyi0Wz67DN15W3girJ5kyB5Zb7vk0LK6r7gbQ/) on customer specific keys.  [Planning process changes](https://github.com/sourcegraph/about/pull/1449) were merged, along with changes to the [sample problem](https://docs.google.com/document/d/1oXhjU_3y2uhpmWJ2hD0NTaqSanUAOlKDFFmjazAU3Rg/).  I also cleared the decks with Lever candidates for our [security position](https://about.sourcegraph.com/handbook/engineering/hiring/software-engineer-security), and got our tester across the finish line.

**Next Week**

This week, I'm going to support our new tester, so that he can get going with async testing of the application.  @daxmc99 and I will continue encrypting data during initialization, with the stretch of supporting rotating keys as well.  Finally, I'd like to at least start a blog post to share the results of [security issue #83](https://github.com/sourcegraph/security-issues/issues/83) and talk about transparent security.

**Comment by ElizabethStirling** [comment]
**This Week**
Closed out #1126 and [Security Issue 73](https://github.com/sourcegraph/security-issues/issues/73), both of which revolved around invalidation of a user's sessions. User sessions are now invalidated when the user signs out of one browser, as well as when the user's password is changed. Site admins now have a way to manually invalidate sessions belonging to a user. I didn't need to split things up into two separate PRs as I had initially feared. I took a look at #7517, which entails sending a user an email when important information is updated. However, due to reprioritization, I'm going to focus on [Security Issue 74](https://github.com/sourcegraph/security-issues/issues/74), which allows referrers to acquire password reset links.

**Next Week**
I plan to work on [Security Issue 74](https://github.com/sourcegraph/security-issues/issues/74) and #7517. As mentioned above, these will deal with password reset security and informing the user when important account information changes. I hope to be able to get both of these in late stage PRs or merged by Thursday, so that, as a stretch goal, I can try to get a start on #10629. I talked with @efritz and @eseliger about this, and will hopefully follow up with them on how to best debug and fix this issue.

**Comment by sushiwushi** [comment]
**This Week**
Spent my time around to build Sourcegraph via [docker](https://docs.sourcegraph.com/admin/install/docker) locally. Still pretty new to everything so I'm taking my time to make sure I didn't left anything out.

**Next Week**
I plan to start analyzing available roles to test for privilege escalation vulnerabilities. (Eg. A regular member might be able to perform admin actions). I will make use of [Burp Suite](https://portswigger.net/burp) with the [Autorize plugin](https://portswigger.net/bappstore/f9bbac8c4acf4aefa4d7dc92a991af2f) to test the application.

**Comment by chayim** [comment]
**This Week**

This week was mostly focused work on the 3.20 release.  We made a tonne of progress on encrypting data during initialization, and even automatic key rotation! There are mostly [PR comments](https://github.com/sourcegraph/sourcegraph/pull/13080/) left and not new development, as part of the 3.20 release.  I interviewed someone as well, and both reviewed and shared internally the results of [security issue #83](https://github.com/sourcegraph/security-issues/issues/83).  The team also expanded by getting our [internal tester](https://github.com/sushiwushi) going, and an interview was conducted.

**Next Week**

I'm planning on finishing up the 3.20 release, whilst we plan for 3.21. I'll be continuing to interview candidates, and work on goals. Finally, I'd like to at least start on the blog post mentioned [previously](https://github.com/sourcegraph/sourcegraph/issues/13130#issuecomment-688077499).  Due to the pandemic, my availability will be reduced, but it's unknown as of yet, to what degree.

**Comment by davejrt** [comment]
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.20 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

**Comment by ElizabethStirling** [comment]
**This week**
I started this week getting [Security Issue 74](https://github.com/sourcegraph/security-issues/issues/74) hammered out. I did some investigation into #7517, and have a good idea how I'll go about implementing it. After that, I ran a technical experience interview for a candidate for another security software engineer. I spent the rest of the week answering some security questions from design and sales, as well as proposing a change to the company values to add [High Agency](https://github.com/sourcegraph/about/pull/1603).

**Next week**
I hope to get the aforementioned PR on [our Company Values](https://github.com/sourcegraph/about/pull/1603). Concurrently, I'll get #7517 hammered out, and get a good start on #10629. Additionally, I'll be out on Friday the 25th for my mandatory day off this month.

**Comment by sushiwushi** [comment]
**This Week**

Spent most of the time analyzing the surface area of Sourcegraph since my aim is to find as many low hanging fruits as possible. So far, I discovered a bug which can be found [here](https://github.com/sourcegraph/security-issues/issues/94).

**Next Week**

I'm planning to focus and determine the highest potential impact that can be abused through the vulnerability. This will allow people to have a more understanding on what the vulnerability can cause if exploited in wild. If time permits, I will continue to immerse myself in testing the application.

**Comment by chayim** [comment]
**Last Week**

Continued to work towards the 3.20 release, and plan for 3.21. An interview candidate was moved forward in the funnel, whilst a PR was issued for goals.  I started on the [blog post](https://docs.google.com/document/d/12joS1U3oh-wK-UvN2sIdVULZA-mtJqknUmb3jlaoh1g) mentioned previously.

**This week**

This week is shorter as I was off the 20th for RH. The main focus on the upcoming week is team building, goals, and 3.21 planning. I'll continue to support 3.20 by documenting security artifacts such as our data policies.  Finally, I'd like to complete [the blog post](https://docs.google.com/document/d/12joS1U3oh-wK-UvN2sIdVULZA-mtJqknUmb3jlaoh1g) started last week.
